crates/bevy_ui/src/render/ui_material_pipeline.rs

```diff
@@ -0,0 +1,774 @@
+use std::{hash::Hash, marker::PhantomData, ops::Range};
+
+use bevy_app::{App, Plugin};
+use bevy_asset::{load_internal_asset, AssetApp, AssetEvent, AssetId, AssetServer, Assets, Handle};
+use bevy_derive::{Deref, DerefMut};
+use bevy_ecs::{
+    prelude::{Component, Entity, EventReader},
+    query::{ROQueryItem, With},
+    schedule::IntoSystemConfigs,
+    storage::SparseSet,
+    system::{
+        lifetimeless::{Read, SRes},
+        Commands, Local, Query, Res, ResMut, Resource, SystemParamItem,
+    },
+    world::{FromWorld, World},
+};
+use bevy_math::{Mat4, Rect, Vec2, Vec4Swizzles};
+use bevy_render::{
+    extract_component::ExtractComponentPlugin,
+    render_asset::RenderAssets,
+    render_phase::{
+        AddRenderCommand, DrawFunctions, PhaseItem, RenderCommand, RenderCommandResult,
+        RenderPhase, SetItemPipeline, TrackedRenderPass,
+    },
+    render_resource::{
+        AsBindGroupError, BindGroup, BindGroupEntries, BindGroupLayout, BindGroupLayoutDescriptor,
+        BindGroupLayoutEntry, BindingType, BlendState, BufferBindingType, BufferUsages, BufferVec,
+        ColorTargetState, ColorWrites, FragmentState, FrontFace, MultisampleState,
+        OwnedBindingResource, PipelineCache, PolygonMode, PrimitiveState, PrimitiveTopology,
+        RenderPipelineDescriptor, Shader, ShaderRef, ShaderStages, ShaderType,
+        SpecializedRenderPipeline, SpecializedRenderPipelines, TextureFormat, VertexBufferLayout,
+        VertexFormat, VertexState, VertexStepMode,
+    },
+    renderer::{RenderDevice, RenderQueue},
+    texture::{BevyDefault, FallbackImage, Image},
+    view::{
+        ExtractedView, ViewTarget, ViewUniform, ViewUniformOffset, ViewUniforms, ViewVisibility,
+    },
+    Extract, ExtractSchedule, Render, RenderApp, RenderSet,
+};
+use bevy_transform::prelude::GlobalTransform;
+use bevy_utils::{FloatOrd, HashMap, HashSet};
+use bevy_window::{PrimaryWindow, Window};
+use bytemuck::{Pod, Zeroable};
+
+use crate::{
+    resolve_border_thickness, CalculatedClip, Node, RenderUiSystem, Style, TransparentUi,
+    UiMaterial, UiMaterialKey, UiScale, UiStack, QUAD_INDICES, QUAD_VERTEX_POSITIONS,
+};
+
+pub const UI_MATERIAL_SHADER_HANDLE: Handle<Shader> = Handle::weak_from_u128(10074188772096983955);
+
+const UI_VERTEX_OUTPUT_SHADER_HANDLE: Handle<Shader> = Handle::weak_from_u128(10123618247720234751);
+
+/// Adds the necessary ECS resources and render logic to enable rendering entities using the given
+/// [`UiMaterial`] asset type (which includes [`UiMaterial`] types).
```

Comment:
The last part of this sentence is confusing to me.

Reply:
I removed this part as it is confusing and doesn't convey any extra information. (I copied this from the Mesh2d Material pipeline. 👍